java-spiffe-core/src/main/java/io/spiffe/workloadapi/WorkloadApiClient.java

```diff
@@ -0,0 +1,379 @@
+        this.cancellableContexts = Collections.synchronizedList(new ArrayList<>());
+        this.backoffPolicy = backoffPolicy;
+        this.retryExecutor = retryExecutor;
+        this.executorService = executorService;
+    }
+
+    /**
+     * Constructor
+     * <p>
+     * It is recommended to use the facility methods `newClient()` to get a new instance
+     * of the WorkloadApiClient.
+     *
+     * @param workloadApiAsyncStub    a {@link SpiffeWorkloadAPIStub}
+     * @param workloadApiBlockingStub a {@link SpiffeWorkloadAPIBlockingStub}
+     * @param managedChannel          a {@link ManagedChannelWrapper}
+     */
+    public WorkloadApiClient(final SpiffeWorkloadAPIStub workloadApiAsyncStub,
+                             final SpiffeWorkloadAPIBlockingStub workloadApiBlockingStub,
+                             final ManagedChannelWrapper managedChannel) {
+        this.workloadApiAsyncStub = workloadApiAsyncStub;
+        this.workloadApiBlockingStub = workloadApiBlockingStub;
+        this.backoffPolicy = new BackoffPolicy();
+        this.executorService = Executors.newCachedThreadPool();
+        this.retryExecutor = Executors.newSingleThreadScheduledExecutor();
+        this.cancellableContexts = Collections.synchronizedList(new ArrayList<>());
+        this.managedChannel = managedChannel;
+    }
+
+    /**
+     * Creates a new Workload API client using the default socket endpoint address.
+     *
+     * @return a {@link WorkloadApiClient}
+     * @throws SocketEndpointAddressException if the Workload API socket endpoint address is not valid
+     * @see Address#getDefaultAddress()
+     */
+    public static WorkloadApiClient newClient() throws SocketEndpointAddressException {
+        val options = ClientOptions.builder().build();
+        return newClient(options);
+    }
+
+    /**
+     * Creates a new Workload API client configured with the given client options.
+     * <p>
+     * If the SPIFFE socket endpoint address is not provided in the options, it uses the default address.
+     *
+     * @param options {@link ClientOptions}
+     * @return a {@link WorkloadApiClient}
+     * @throws SocketEndpointAddressException if the Workload API socket endpoint address is not valid
+     */
+    public static WorkloadApiClient newClient(@NonNull final ClientOptions options)
+            throws SocketEndpointAddressException {
+        final String spiffeSocketPath;
+        if (StringUtils.isNotBlank(options.spiffeSocketPath)) {
+            spiffeSocketPath = options.spiffeSocketPath;
+        } else {
+            spiffeSocketPath = Address.getDefaultAddress();
+        }
+
+        if (options.backoffPolicy == null) {
+            options.backoffPolicy = new BackoffPolicy();
+        }
+
+        if (options.executorService == null) {
+            options.executorService = Executors.newCachedThreadPool();
+        }
+
+        val socketEndpointAddress = Address.parseAddress(spiffeSocketPath);
```

Comment:
parsing and validating the address at runtime is unavoidable, because "default address" means that the address path is read from the env variable SPIFFE_ENDPOINT_SOCKET, so we still need to parse and validate the value set in that variable.

Reply:
Got it, makes sense then.